bench/workgen/workgen.cpp

```diff
@@ -1367,90 +1529,132 @@ ThreadRunner::op_run(Operation *op)
+        std::map<std::string, tint_t>::iterator itr;
+        uint32_t retries = 0;
+        size_t num_tables;
+
+        while (
+          (num_tables = _icontext->_dyn_table_names.size()) > 0 && ++retries < TABLE_MAX_RETRIES) {
+            itr = _icontext->_dyn_tint.begin();
+            std::advance(itr, random_value() % num_tables);
+
+            if (_icontext->_dyn_table_runtime[itr->second]._is_base &&
+              !_icontext->_dyn_table_runtime[itr->second]._pending_delete) {
+                break;
+            }
         }
-    } else {
-        tint = op->_table._internal->_tint;
-        table_uri = op->_table._uri;
+        if (num_tables == 0 || retries >= TABLE_MAX_RETRIES) { // Try again next time.
+            return 0;
+        }
+
+        std::string op_uri = itr->first; // Get the table name.
+        tint_t op_tint = itr->second;    // Get the tint.
+        op_kv_gen(op, op_tint);          // Set the key and value for the operation.
+
+        // Use atomic here as we can race with another thread that acquires the shared lock.
+        (void)workgen_atomic_add32(&_icontext->_dyn_table_runtime[op_tint]._in_use, 1);
+
+        // Do we need to mirror operations? If not, we are done here.
+        if (!_icontext->_dyn_table_runtime[op_tint].has_mirror()) {
+            op_set_table(op, op_uri, op_tint);
+            return op_run(op);
+        }
+
+        // Copy this operation to two new operations on the base table and the mirror.
+        base_op = *op;
+        op_set_table(&base_op, op_uri, op_tint);
+
+        mirror_op = *op;
+        std::string mirror_op_uri = _icontext->_dyn_table_runtime[op_tint]._mirror;
+        tint_t mirror_op_tint = _icontext->_dyn_tint[mirror_op_uri];
+        op_set_table(&mirror_op, mirror_op_uri, mirror_op_tint);
+        (void)workgen_atomic_add32(&_icontext->_dyn_table_runtime[mirror_op_tint]._in_use, 1);
+        ASSERT(!_icontext->_dyn_table_runtime[mirror_op_tint]._pending_delete);
     }
 
-    /*
-     * A potential race: thread1 is inserting, and increments Context->_recno[] for fileX.wt.
-     * thread2 is doing one of remove/search/update and grabs the new value of Context->_recno[] for
-     * fileX.wt. thread2 randomly chooses the highest recno (which has not yet been inserted by
-     * thread1), and when it accesses the record will get WT_NOTFOUND. It should be somewhat rare
-     * (and most likely when the threads are first beginning). Any WT_NOTFOUND returns are allowed
-     * and get their own statistic bumped.
-     */
+    // Create a new operation with the base and mirror table ops. Group them in a transaction
+    // unless we are already in a transaction.
+    Operation new_op;
+    std::vector<Operation> new_group = {base_op, mirror_op};
+    new_op._group = &new_group;
+    new_op._repeatgroup = 1;
+
+    if (!_in_transaction) {
+        Transaction txn;
+        new_op.transaction = &txn;
+    }
+
+    return op_run(&new_op);
```

Comment:
I love this code!

Except for one thing.  I think when you have:
```
 if (!_in_transaction) {
        Transaction txn;
        new_op.transaction = &txn;
    }
```
txn is on the stack, and when we exit the `}`, isn't its address invalid?  So I think Transction txn must be declared outside the `if` clause (and if it's not used, that's okay).

Reply:
Good catch - thanks! Fixed in [6c72925](https://github.com/wiredtiger/wiredtiger/pull/8798/commits/6c7292568c1e15f405d3e32eec637aca45361cdc).